contracts/shapella-upgrade/ShapellaUpgradeTemplate.sol

```diff
@@ -0,0 +1,506 @@
+        _finishUpgrade();
+    }
+
+    /// Perform basic checks to revert the entire upgrade if something gone wrong
+    function verifyFinishedUpgrade() external view {
+        _verifyFinishedUpgrade();
+    }
+
+    function _startUpgrade() internal {
+        if (msg.sender != _voting) revert OnlyVotingCanUpgrade();
+        if (isUpgradeStarted) revert CanOnlyStartOnce();
+        isUpgradeStarted = true;
+
+        // Need to check / set locator implementation first because the other initial state checks depend on correct locator
+        if (IOssifiableProxy(address(_locator)).proxy__getImplementation() != _locatorImplementation) {
+            IOssifiableProxy(address(_locator)).proxy__upgradeTo(_locatorImplementation);
+        }
+
+        _verifyInitialState();
+
+        _prepareAccountingOracle();
+
+        _prepareValidatorsExitBus();
+
+        _migrateLidoOracleCommitteeMembers();
+
+        _prepareWithdrawalQueue();
+
+        _prepareStakingRouter();
+
+        _migrateDSMGuardians();
+    }
+
+    function _verifyInitialState() internal view {
+        if (ILidoOracle(_locator.legacyOracle()).getVersion() != 3) revert LidoOracleMustNotBeUpgradedToLegacyYet();
+
+        _verifyProxyAdmins(address(this));
+
+        // TODO: put back if enough gas
+        _verifyInitialProxyImplementations();
+
+        _verifyOZAdmins(address(this));
+
+        if (IDepositSecurityModule(_locator.depositSecurityModule()).getOwner() != address(this)) revert WrongDsmOwner();
+    }
+
+    function _verifyProxyAdmins(address admin) internal view {
+        if (IOssifiableProxy(address(_locator)).proxy__getAdmin() != admin) revert WrongLocatorAdmin();
+        if (IOssifiableProxy(_locator.withdrawalQueue()).proxy__getAdmin() != admin) revert WrongWQAdmin();
+        if (IOssifiableProxy(_locator.stakingRouter()).proxy__getAdmin() != admin) revert WrongSQAdmin();
+        if (IOssifiableProxy(_locator.validatorsExitBusOracle()).proxy__getAdmin() != admin) revert WrongEBAdmin();
+        if (IOssifiableProxy(_locator.accountingOracle()).proxy__getAdmin() != admin) revert WrongAOAdmin();
+    }
+
+    function _verifyInitialProxyImplementations() internal view {
+        if (IOssifiableProxy(_locator.withdrawalQueue()).proxy__getImplementation() != _dummyImplementation) revert WrongWQInitialImpl();
+        if (IOssifiableProxy(_locator.stakingRouter()).proxy__getImplementation() != _dummyImplementation) revert WrongSRInitialImpl();
+        if (IOssifiableProxy(_locator.validatorsExitBusOracle()).proxy__getImplementation() != _dummyImplementation) revert WrongEBInitialImpl();
+        if (IOssifiableProxy(_locator.accountingOracle()).proxy__getImplementation() != _dummyImplementation) revert WrongAOInitialImpl();
+    }
+
+    function _verifyOZAdmins(address admin) internal view {
+        if (IAccessControlEnumerable(_hashConsensusForAccountingOracle).getRoleMemberCount(DEFAULT_ADMIN_ROLE) != 1) revert MultipleAdminsHCAO();
```

Comment:
can we define just the single 'MultipleAdmins(string name, address contract)` error?